File: learn-algo/leetcode/689/c++/689.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> maxSumOfThreeSubarrays(vector<int> &nums, int k) {
        int n = nums.size();
        vector<int> s(n + 1);
        for (auto i = 1; i <= n; ++i) {
            s[i] = s[i - 1] + nums[i - 1];
        }

        int x = n + 1, y = 3;
        vector<vector<int>> f(n + 2, vector<int>(4));
        for (int i = n - k + 1; i > 0; --i) {
            for (int j = 1; j <= 3; ++j) {
                f[i][j] = max(f[i + 1][j],
                              f[i + k][j - 1] + s[i + k - 1] - s[i - 1]); // s[i+n+1] - s[i] 表示下标从i到i+n的所有数的和
            }
            if (f[x][3] <= f[i][3]) {
                x = i;
            }
        }

        vector<int> res;
        while (y > 0) {
            while (f[x][y] != f[x + k][y - 1] + s[x + k - 1] - s[x - 1]) { // x到x+k-1的所有的和
                x++;
            }

            res.push_back(x - 1);
            x += k;
            y--;
        }

        return res;
    }
};

int main() {
    Solution s;
    vector<int> in = {1, 2, 1, 2, 6, 7, 5, 1};
    int k = 2;
    auto res = s.maxSumOfThreeSubarrays(in, k);
    for_each(res.begin(), res.end(), [](auto t) { cout << t << " "; });
    cout << endl;
    return 0;
}

File: learn/learn-algo/leetcode/689/go/689.go
package main

import "fmt"

func maxSumOfThreeSubarrays(nums []int, k int) []int {
	n := len(nums)
	var s = make([]int, n+1)
	for i := 1; i <= n; i++ {
		s[i] = s[i-1] + nums[i-1]
	}

	var x, y = n + 1, 3
	var f = make([][4]int, n+2)

	for i := n - k + 1; i > 0; i-- {
		for j := 1; j <= 3; j++ {
			f[i][j] = max(f[i+1][j], f[i+k][j-1]+s[i-1+k]-s[i-1])
		}

		if f[x][3] <= f[i][3] {
			x = i
		}
	}

	var res []int
	for y > 0 {
		for f[x][y] != f[x+k][y-1]+s[x+k-1]-s[x-1] {
			x++
		}

		res = append(res, x-1)
		y--
		x += k
	}
	return res
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func main() {
	var in = []int{1, 2, 1, 2, 6, 7, 5, 1}
	var res = maxSumOfThreeSubarrays(in, 2)
	fmt.Println(res)
}
